### PR TITLE
Allow customizing session closure

### LINKED ISSUE #1: Allow customization of session closure
The default logic, commit on success, rollback on failure is a sensible one
https://github.com/vytas7/falcon-sqla/blob/e4d7865892c1111c49ecb7e06a7b2a425990ec73/falcon_sqla/middleware.py#L53-L56

maybe it could be useful to add the ability of a resource to customize this logic.
I'm not sure how to best implement it. Probably a function in the resource class that returns what to do (commit, rollback, nothing) is the best solution.

### PR DESCRIPTION
**Pull request by vytas7**
Closes #1

### CODE REVIEW

File: falcon_sqla/constants.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    When :attr:`choosing <falcon_sqla.Manager.get_bind>`, this
+    engine will participate in balancing load in both :attr:`~EngineRole.READ`
+    and :attr:`~EngineRole.WRITE` contexts (unless
+    :attr:`~.SessionOptions.read_from_rw_engines` is set to ``True``).
+    """
+
+
+class SessionCleanup(enum.Enum):
+    """
+    Session cleanup behavior.
```
Comment (CaselIT) [inline falcon_sqla/constants.py]:
```suggestion
    """Session cleanup behavior.
```

File: falcon_sqla/manager.py [hunk 2]
```diff
@@ -122,6 +136,29 @@ def get_session(self, req=None, resp=None):
...
+            if (session_cleanup == COMMIT_ON_SUCCESS and succeeded
+                    or session_cleanup == COMMIT):
+                session.commit()
+            elif session_cleanup != CLOSE_ONLY:
+                session.rollback()
+        finally:
+            if req and resp:
+                # NOTE(vytas): Break circular references between the request
+                #   and session in case the latter was stored in req.context.
+                del session.info['req']
```
Comment (CaselIT) [inline falcon_sqla/manager.py:162]: not sure if it makes sense to also do `req.context.session = None`

File: tests/test_session_scope.py [hunk 3]
```diff
@@ -71,3 +72,26 @@ def test_generic_scope(database):
         malbolge = session.query(database.Language).first()
         assert malbolge.name == 'Malbolge'
         assert malbolge.created == 1998
+
+
+@pytest.mark.parametrize('cleanup,expected', [
+    (SessionCleanup.CLOSE_ONLY, 1999),
+    (SessionCleanup.COMMIT, 1998),
+    (SessionCleanup.COMMIT_ON_SUCCESS, 1998),
```
Comment (CaselIT) [inline tests/test_session_scope.py:81]: we are not testing that it does not commit on error

**Comment by vytas7** [comment]
Yeah, thanks, this is still in semi-WiP 🙂
The biggest concern atm is that something is hanging with Postgres, will investigate.

**Comment by vytas7** [comment]
(Got it, it was a problem with the tests, will fix later.)

File: falcon_sqla/manager.py (hunk 2 above)
Comment (vytas7) [reply falcon_sqla/manager.py:162]: Dunno, it's hopefully not needed really, as long as we delete all the references from the session's context, there should be no circular dependencies there (Python can handle these too, but it usually takes more time).

File: tests/test_session_scope.py (hunk 3 above)
Comment (vytas7) [reply tests/test_session_scope.py:81]: I need to rework the tests of the package wholesale, but I've now added a test for that.
Not sure if it makes sense to have that "commit" mode that tries to commit regardless?

**Pull request merged**